dbms/src/Common/FieldVisitors.h

```diff
@@ -492,4 +414,48 @@ class FieldVisitorSum : public StaticVisitor<bool>
     }
 };
 
+/**
+  * castField() implementation details follow.
+  */
+template<typename DestType, typename SrcType>
+DestType staticCastFieldValue(const SrcType & src)
+{
+    auto dest = static_cast<DestType>(src);
+    if (!accurate::equalsOp(dest, src))
+    {
+        throw Exception("Cannot cast Field value '" + applyVisitor(FieldVisitorToString(), Field(src))
+                        + "' of type '" + demangle(typeid(SrcType).name())
+                        + "' to '" + demangle(typeid(DestType).name()) + "'",
+                        ErrorCodes::VALUE_IS_OUT_OF_RANGE_OF_DATA_TYPE);
+    }
+    return dest;
+}
+
+template<typename DestType, typename SrcType>
+DestType castFieldValue(const SrcType & src)
+{
+    if constexpr (std::is_convertible_v<DestType, SrcType>)
+    {
+        return staticCastFieldValue<DestType, SrcType>(src);
+    }
+    else
+    {
+        throw Exception("Cannot cast Field value of type '" + demangle(typeid(SrcType).name())
+                        + "' to '" + demangle(typeid(DestType).name()) + "'",
+                        ErrorCodes::LOGICAL_ERROR);
```

Comment:
Should throw TYPES_MISMATCH, because it processes user-supplied data such as the AST.